Adding LAPACK implementation of LU Factorization

CaptainSharf opened the pull request:
As part of this PR
- Adding procedure `lu`
- Adding tests to verify the correctness of `lu`

ben-albrecht commented on modules/packages/LinearAlgebra.chpl:
```diff
@@ -1313,6 +1318,28 @@ proc lu (A: [?Adom] ?eltType) {
+  
+  use SysCTypes;
+  use LAPACK;
+  var A_clone = A;
```
Chapel tends to use `camelCase` over `snake_case`. Maybe call this `Acopy`?

ben-albrecht commented on modules/packages/LinearAlgebra.chpl:
```diff
@@ -1313,6 +1318,28 @@ proc lu (A: [?Adom] ?eltType) {
+
+  var rows = Adom.shape[0];
+  var cols = Adom.shape[1];
+  var ipiv: [{1..rows}] c_int;
```
You can plug a range directly into the array declaration:

```suggestion
  var ipiv: [1..rows] c_int;
```

ben-albrecht commented on modules/packages/LinearAlgebra.chpl:
```diff
@@ -1313,6 +1318,28 @@ proc lu (A: [?Adom] ?eltType) {
+  use LAPACK;
+  var A_clone = A;
+
+  var rows = Adom.shape[0];
```
This can also be:

```chpl
const (rows, cols) = Adom.shape;
```

ben-albrecht commented on modules/packages/LinearAlgebra.chpl:
```diff
@@ -1313,6 +1318,28 @@ proc lu (A: [?Adom] ?eltType) {
+  var rows = Adom.shape[0];
+  var cols = Adom.shape[1];
+  var ipiv: [{1..rows}] c_int;
+  var info = getrf(lapack_memory_order.row_major,A_clone,ipiv);
```
```suggestion
  var info = getrf(lapack_memory_order.row_major, A_clone, ipiv);
```

ben-albrecht commented on modules/packages/LinearAlgebra.chpl:
```diff
@@ -1313,6 +1318,28 @@ proc lu (A: [?Adom] ?eltType) {
+  var ipiv: [{1..rows}] c_int;
+  var info = getrf(lapack_memory_order.row_major,A_clone,ipiv);
+  var U = triu(A_clone);
+  var L = A_clone-U+eye(A.domain);
```
```suggestion
  var L = A_clone - U + eye(A.domain);
```

ben-albrecht commented on modules/packages/LinearAlgebra.chpl:
```diff
@@ -1313,6 +1318,28 @@ proc lu (A: [?Adom] ?eltType) {
+  var info = getrf(lapack_memory_order.row_major,A_clone,ipiv);
+  var U = triu(A_clone);
+  var L = A_clone-U+eye(A.domain);
+  return(L[1..rows,1..rows],U,ipiv);
```
This return type differs from the other overload, right? I don't think that will work as is, so we'll either need to determine that we are changing the return type in https://github.com/chapel-lang/chapel/issues/16973 or combine `L` and `U` here.

ben-albrecht reviewed (commented):
I don't see a new test yet, so I assume this is still work-in-progress. I've provided some early feedback to help you along the way though. 

Thanks for the contribution!

CaptainSharf commented:
Hi @ben-albrecht ,
> I don't see a new test yet, so I assume this is still work-in-progress. I've provided some early feedback to help you along the way though.
Yes this is still work in progress

ben-albrecht commented:
@CaptainSharf - I am revisiting some old open PRs. I was going to close this PR due to inactivity, but saw you recently referenced it. Do you plan to revisit it sometime soon? If not, I suggest we close it until you have time to resume in the future.

CaptainSharf commented:
Hi @ben-albrecht ,
Thanks for your review. I added tests for the LU decomposition as part of PR #16980 . Would it be OK if I added the same tests here?

CaptainSharf commented:
I wish to work on this feature in the next week or so. I believe that the LAPACK implementation of LU factorization is stable and might provide us with a useful workaround until we fix the issues with the native version of the LU factorization.

ben-albrecht reviewed (changes requested):
Can you copy the existing LU tests up into `test/library/packages/LinearAlgebra/correctness/testLU.{chpl, good}` so that we can test lu with LAPACK enabled?
